File: Tests.Csharp/Connecting.cs
using System.Security.Cryptography.X509Certificates;
using UmaDb.Client;
using Xunit;

namespace Tests.Csharp;

public class Connecting
{
    [Fact]
    public void Connect_throws_on_null_options()
    {
        var ex = Assert.Throws<ArgumentNullException>(() => UmaClient.Connect(null!));
        Assert.Equal("options", ex.ParamName);
    }

    [Fact]
    public void Connect_throws_on_empty_host()
    {
        var ex = Assert.Throws<ArgumentException>(() =>
            UmaClient.Connect(new UmaClientOptions().WithHost("").WithPort(50051)));
        Assert.Equal("options", ex.ParamName);
    }

    [Fact]
    public void Connect_throws_on_invalid_port()
    {
        var ex = Assert.Throws<ArgumentOutOfRangeException>(() =>
            UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(0)));
        Assert.Equal("options", ex.ParamName);
    }

    [Fact]
    public async Task can_create_uma_client_to_http_server()
    {
        using var umaClient = UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50051));
        await umaClient.GetHeadAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task cannot_connect_with_tls_to_http_only_server()
    {
        using var umaClient =
            UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50051).EnableTls());
        await Assert.ThrowsAnyAsync<UmaDbException>(() =>
            umaClient.GetHeadAsync(TestContext.Current.CancellationToken).AsTask());
    }

    [Fact]
    public async Task cannot_connect_with_api_key_to_http_only_server()
    {
        using var umaClient =
            UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50051).WithApiKey("key"));
        await Assert.ThrowsAnyAsync<UmaDbException>(() =>
            umaClient.GetHeadAsync(TestContext.Current.CancellationToken).AsTask());
    }

    [Fact]
    public async Task cannot_create_uma_client_without_tls_if_server_requires_tls()
    {
        using var umaClient = UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50001));

        await Assert.ThrowsAnyAsync<UmaDbException>(() =>
            umaClient.GetHeadAsync(TestContext.Current.CancellationToken).AsTask());
    }

    [Fact]
    public async Task cannot_create_uma_client_without_apikey_if_server_requires_tls_with_apikey()
    {
        using var umaClient =
            UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50001).WithCaCert("certs/ca.pem"));

        var exception = await Assert.ThrowsAsync<UmaDbException.AuthenticationException>(() =>
            umaClient.GetHeadAsync(TestContext.Current.CancellationToken).AsTask());
        Assert.IsType<UmaDbException.AuthenticationException>(exception);
        Assert.Equal("authentication error: missing or invalid API key", exception.Message);
    }

    [Fact]
    public async Task cannot_create_uma_client_with_wrong_apikey()
    {
        using var umaClient = UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50001)
            .WithCaCert("certs/ca.pem").WithApiKey("wrong-api-key"));

        var exception = await Assert.ThrowsAsync<UmaDbException.AuthenticationException>(() =>
            umaClient.GetHeadAsync(TestContext.Current.CancellationToken).AsTask());
        Assert.IsType<UmaDbException.AuthenticationException>(exception);
        Assert.Equal("authentication error: missing or invalid API key", exception.Message);
    }

    [Fact]
    public async Task can_create_uma_client_to_tls_server_with_api_key()
    {
        using var umaClient = UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50001)
            .WithCaCert("certs/ca.pem").WithApiKey("test-api-key"));
        await umaClient.GetHeadAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task can_connect_with_well_known_ca_and_api_key()
    {
        var caPath = Path.Combine(AppContext.BaseDirectory, "certs", "ca.pem");
        if (!File.Exists(caPath))
            throw new InvalidOperationException($"Test CA not found at {caPath}. Ensure certs are copied to output.");

        using var caCert = X509CertificateLoader.LoadCertificateFromFile(caPath);
        using (var store = new X509Store(StoreName.Root, StoreLocation.CurrentUser))
        {
            store.Open(OpenFlags.ReadWrite);
            store.Add(caCert);
            try
            {
                using var umaClient = UmaClient.Connect(new UmaClientOptions()
                    .WithHost("localhost")
                    .WithPort(50001)
                    .WithApiKey("test-api-key")
                    .EnableTls());
                await umaClient.GetHeadAsync(TestContext.Current.CancellationToken);
            }
            finally
            {
                store.Remove(caCert);
            }
        }
    }
}

File: Tests.Fsharp/Client/Connecting.fs
module Tests.Client.Connecting

open System
open System.IO
open System.Security.Cryptography.X509Certificates
open System.Threading
open System.Threading.Tasks
open FSharp.Control
open Xunit
open UmaDb.Client.ClientBuilder
open UmaDb.Client.Errors
open UmaDb.Client.Operations

[<Fact>]
let ``Connect throws on empty host`` () =
    let ex = Assert.Throws<ArgumentException>(fun () -> connect "" 50051 |> build |> ignore)
    Assert.Equal("host", ex.ParamName)

[<Fact>]
let ``Connect throws on invalid port`` () =
    let ex = Assert.Throws<ArgumentException>(fun () -> connect "localhost" 0 |> build |> ignore)
    Assert.Equal("port", ex.ParamName)


[<Fact>]
let ``can create uma client to http server`` () =
    task {
        use umaClient = connect "localhost" 50002 |> build
        let! _ = readHead umaClient TestContext.Current.CancellationToken
        ()
    }

[<Fact>]
let ``cannot_connect_with_tls_to_http_only_server`` () =
    (fun () ->
        task {
            use client = connect "localhost" 50002 |> withTls |> build
            let! _ = readHead client TestContext.Current.CancellationToken
            ()
        }
        :> Task)
    |> Assert.ThrowsAnyAsync<UmaDbException>

[<Fact>]
let ``cannot connect with api key to http only server`` () =
    (fun () ->
        task {
            use umaClient = connect "localhost" 50002 |> withApiKey "key" |> build
            let! _ = readHead umaClient TestContext.Current.CancellationToken
            ()
        }
        :> Task)
    |> Assert.ThrowsAnyAsync<UmaDbException>

[<Fact>]
let ``cannot create uma client without tls if server requires tls`` () =
    task {
        let work () =
            task {
                use umaClient = connect "localhost" 50003 |> build
                let! _ = readHead umaClient TestContext.Current.CancellationToken
                ()
            }
            :> Task
        let! _ex = Assert.ThrowsAsync<UmaDbException>(work)
        ()
    }

[<Fact>]
let ``cannot create uma client without apikey if server requires tls with apikey`` () =
    task {
        let work () =
            task {
                use umaClient = connect "localhost" 50003 |> withCaCert "certs/ca.pem" |> build
                let! _ = readHead umaClient TestContext.Current.CancellationToken
                ()
            }
            :> Task
        let! ex = Assert.ThrowsAsync<AuthenticationException>(work)
        Assert.Equal("Authentication error: missing or invalid API key", ex.Message)
    }

[<Fact>]
let ``cannot create uma client with wrong apikey`` () =
    task {
        let work () =
            task {
                use umaClient = connect "localhost" 50003 |> withCaCert "certs/ca.pem" |> withApiKey "wrong-api-key" |> build
                let! _ = readHead umaClient TestContext.Current.CancellationToken
                ()
            }
            :> Task
        let! ex = Assert.ThrowsAsync<AuthenticationException>(work)
        Assert.Equal("Authentication error: missing or invalid API key", ex.Message)
    }

[<Fact>]
let ``can create uma client to tls server with api key`` () =
    task {
        use umaClient = connect "localhost" 50003 |> withCaCert "certs/ca.pem" |> withApiKey "test-api-key" |> build
        let! _ = readHead umaClient CancellationToken.None
        ()
    }

[<Fact>]
let ``can connect with well known ca and api key`` () =
    task {
        let caPath = Path.Combine(AppContext.BaseDirectory, "certs", "ca.pem")
        if not (File.Exists(caPath)) then
            raise (InvalidOperationException($"Test CA not found at {caPath}. Ensure certs are copied to output."))
        let certPem = File.ReadAllText(caPath)
        use caCert = X509Certificate2.CreateFromPem(certPem)
        use store = new X509Store(StoreName.Root, StoreLocation.CurrentUser)
        store.Open OpenFlags.ReadWrite
        store.Add caCert
        try
            use umaClient = connect "localhost" 50003 |> withTls |> withApiKey "test-api-key" |> build
            let! _ = readHead umaClient CancellationToken.None
            ()
        finally
            store.Remove(caCert)
    }
